wgpu-hal/src/gles/queue.rs

```diff
@@ -214,21 +217,58 @@ impl super::Queue {
             }
             C::CopyBufferToBuffer {
                 src,
-                src_target,
+                src_target: _,
                 dst,
                 dst_target,
                 copy,
             } => {
-                gl.bind_buffer(src_target, Some(src));
-                gl.bind_buffer(dst_target, Some(dst));
+                let is_index_buffer_only_element_dst = !self
+                    .shared
+                    .private_caps
+                    .contains(super::PrivateCapabilities::INDEX_BUFFER_ROLE_CHANGE)
+                    && dst_target == glow::ELEMENT_ARRAY_BUFFER;
 
-                gl.copy_buffer_sub_data(
-                    src_target,
-                    dst_target,
-                    copy.src_offset as i32,
-                    copy.dst_offset as i32,
-                    copy.size.get() as i32,
-                );
+                let copy_src_target = glow::COPY_READ_BUFFER;
+
+                // WebGL not allowed to copy data from other targets to element buffer and can't copy element data to other buffers
+                let copy_dst_target = if is_index_buffer_only_element_dst {
+                    glow::ELEMENT_ARRAY_BUFFER
+                } else {
+                    glow::COPY_WRITE_BUFFER
+                };
+
+                gl.bind_buffer(copy_src_target, Some(src));
+                gl.bind_buffer(copy_dst_target, Some(dst));
+
+                if is_index_buffer_only_element_dst {
+                    let mut buffer_data = vec![0; copy.size.get() as usize];
+                    gl.get_buffer_sub_data(
```

Comment:
is this necessary? copying buffers should be GPU-local operation, but this code reads the data to CPU

Reply:
I'm not certain. I just grabbed the code from gfx. From the comment above I'm guessing it's copying to the CPU because it's not allowed to copy unmatching buffer types without copying to the CPU first. :man_shrugging: Your guess is better than mine.